File: frontend/src/lib/adminCustomers.ts
// Copyright (c) 2026 Rishi Mule. All Rights Reserved.
// This code and its associated documentation cannot be copied, modified, or distributed without explicit permission from the author.
/**
 * Typed client for the admin customer-supervisor endpoints
 * (see backend/app/src/app/api/admin_customers.py).
 *
 * Every lifecycle write proxied through here records a transition + reason on
 * the backend. The UI collects reason strings (>= 10 chars) before invoking the
 * destructive helpers (suspend / unsuspend / delete / restore).
 */
import { get, post } from "./api";
import type {
  AdminCustomerActivity,
  AdminCustomerHub,
  AdminCustomerListResponse,
  AdminCustomerNotification,
  AdminCustomerOrder,
  CustomerAccountStatus,
  CustomerAddress,
  CustomerLifecycleAction,
} from "@/types";

export interface CustomerListParams {
  q?: string;
  status?: CustomerAccountStatus | "";
  limit: number;
  offset: number;
}

export function fetchCustomerList(token: string, params: CustomerListParams) {
  const sp = new URLSearchParams({
    limit: String(params.limit),
    offset: String(params.offset),
  });
  if (params.q && params.q.trim()) sp.set("q", params.q.trim());
  if (params.status) sp.set("status", params.status);
  return get<AdminCustomerListResponse>(
    `/api/v1/admin/customers?${sp.toString()}`,
    token,
  );
}

export function fetchCustomerHub(customerProfileId: number, token: string) {
  return get<AdminCustomerHub>(
    `/api/v1/admin/customers/${customerProfileId}`,
    token,
  );
}

export function fetchCustomerActivity(
  customerProfileId: number,
  token: string,
) {
  return get<AdminCustomerActivity[]>(
    `/api/v1/admin/customers/${customerProfileId}/activity`,
    token,
  );
}

export function fetchCustomerOrders(customerProfileId: number, token: string) {
  return get<AdminCustomerOrder[]>(
    `/api/v1/admin/customers/${customerProfileId}/orders`,
    token,
  );
}

export function fetchCustomerAddresses(
  customerProfileId: number,
  token: string,
) {
  return get<CustomerAddress[]>(
    `/api/v1/admin/customers/${customerProfileId}/addresses`,
    token,
  );
}

export function fetchCustomerNotifications(
  customerProfileId: number,
  token: string,
) {
  return get<AdminCustomerNotification[]>(
    `/api/v1/admin/customers/${customerProfileId}/notifications`,
    token,
  );
}

/** Apply a lifecycle transition. `reason` must be >= 10 chars (enforced UI-side
 * by AdminReasonModal and backend-side). Resolves to the new status. */
export function customerLifecycleAction(
  customerProfileId: number,
  action: CustomerLifecycleAction,
  reason: string,
  token: string,
) {
  return post<{ status: CustomerAccountStatus }>(
    `/api/v1/admin/customers/${customerProfileId}/${action}`,
    { reason },
    token,
  );
}

File: backend/app/src/app/schemas/admin_customers.py
# Copyright (c) 2026 Rishi Mule. All Rights Reserved.
# This code and its associated documentation cannot be copied, modified, or distributed without explicit permission from the author.
from datetime import datetime

from pydantic import BaseModel, Field


class AdminCustomerActionBody(BaseModel):
    reason: str = Field(min_length=10, max_length=500)


class AdminCustomerListItem(BaseModel):
    customer_profile_id: int
    user_id: int
    email: str
    full_name: str | None
    phone: str | None
    account_status: str
    created_at: datetime


class AdminCustomerList(BaseModel):
    items: list[AdminCustomerListItem]
    total: int


class AdminCustomerHub(BaseModel):
    customer_profile_id: int
    user_id: int
    email: str
    full_name: str | None
    phone: str | None
    account_status: str
    status_reason: str | None
    status_changed_at: datetime | None
    open_orders: int
    open_credit_accounts: int


class AdminCustomerEvent(BaseModel):
    id: int
    actor_user_id: int | None
    actor_role: str
    from_status: str
    to_status: str
    reason: str | None
    created_at: datetime


class AdminCustomerOrder(BaseModel):
    id: int
    store_id: int
    service_name_snapshot: str
    status: str
    total: float
    placed_at: datetime


class AdminCustomerNotification(BaseModel):
    id: int
    type: str
    title: str
    body: str
    read: bool
    created_at: datetime
